File: 03-overrides/node.js/src/data/feature-flag-repository.ts
import db from './database.js';
import { type FeatureFlag, FeatureFlags } from '../types/feature-flag.js';
import type { UserFlag } from '@/types/user-flag.js';


// FRAGILE: value coming back from DB is an int, not a boolean

export default async function getFlag(key: FeatureFlags, name: string | undefined): Promise<boolean> {
  if (!key) {
    throw new Error('Key is required');
  }
  const row = await db.get('SELECT value FROM FeatureFlag WHERE key = ?', [key]);

  let value = row?.value === 1;

  if (name) {
    const userRow = await db.get('SELECT value FROM UserFlag WHERE userId = (SELECT id FROM User WHERE name = ?) AND featureFlagKey = ?', [name, key]);
    if (userRow) {
      value = userRow.value === 1;
    }
  }

  return value;
}

export async function getAllFlags(name: string | undefined): Promise<{ [key: string]: boolean }> {
  const rows: FeatureFlag[] = await db.all('SELECT key, value FROM FeatureFlag', []);
  const results = Object.fromEntries(rows.map(({ key, value }) => [key, !!value]));

  if (name) {
    const userFlags: UserFlag[] = await db.all('SELECT featureFlagKey, value FROM UserFlag WHERE userId = (SELECT id FROM User WHERE name = ?)', [name]);
    for (const { featureFlagKey, value } of userFlags) {
      results[featureFlagKey] = !!value;
    }
  }

  return results;
}

File: 03-overrides/dotnet/Data/FeatureFlagRepository.cs
namespace FeatureFlagWithOverrides.Data;

public interface IFeatureFlagRepository
{
    bool FeatureIsEnabled(FeatureFlags featureFlag, string? name);
    Dictionary<string, bool> GetAllFeatureFlags(string? name);
}

public class FeatureFlagRepository(ApplicationDbContext dbContext) : IFeatureFlagRepository
{
    public bool FeatureIsEnabled(FeatureFlags featureFlag, string? name)
    {
        var flag = dbContext.FeatureFlags.FirstOrDefault(f => f.Key == featureFlag.ToString());
        bool result = flag?.Value ?? false;
        if (!string.IsNullOrEmpty(name))
        {
            var userFlag = dbContext.UserFlags.FirstOrDefault(f => f.FeatureFlagKey == featureFlag.ToString() && f.User.Name == name);
            if (userFlag != null)
            {
                result = userFlag.Value;
            }
        }
        return result;
    }

    public Dictionary<string, bool> GetAllFeatureFlags(string? name)
    {
        var features = dbContext.FeatureFlags.ToList();
        foreach (FeatureFlags flag in Enum.GetValues<FeatureFlags>())
        {
            var feature = features.FirstOrDefault(f => f.Key == flag.ToString());
            if (feature == null)
            {
                feature = new FeatureFlag { Key = flag.ToString(), Value = false };
                features.Add(feature);
            }
        }
        if (!string.IsNullOrEmpty(name))
        {
            var userFlags = dbContext.UserFlags.Where(f => f.User.Name == name).ToList();
            foreach (var userFlag in userFlags)
            {
                var feature = features.FirstOrDefault(f => f.Key == userFlag.FeatureFlagKey);
                if (feature != null)
                {
                    feature.Value = userFlag.Value; // FRAGILE: mutating EntityFramework's cache but intentionally not saving it
                }
                else
                {
                    features.Add(new FeatureFlag { Key = userFlag.FeatureFlagKey, Value = userFlag.Value });
                }
            }
        }
        var results = (
            from f in features
            orderby f.Key
            select f
        ).ToDictionary(f => f.Key, f => f.Value);
        return results;
    }
}
